Allow HTML tags in user Markdown input

mr-vinn opened the pull request:
This change fixes #7066, and issues [#147](https://gitlab.com/gitlab-org/gitlab-ce/issues/147), [#408](https://gitlab.com/gitlab-org/gitlab-ce/issues/408) and [#476](https://gitlab.com/gitlab-org/gitlab-ce/issues/476) from gitlab.com.

The Markdown parser currently uses Redcarpet's `:filter_html` option, which removes all HTML tags from user input.  The parser also calls ActionView's `#sanitize` method, which removes HTML tags and attributes that are not whitelisted.

GitLab's [Markdown documentation](https://github.com/gitlabhq/gitlabhq/blob/master/doc/markdown/markdown.md#inline-html) says that inline HTML should be allowed, which seems incompatible with the `:filter_html` option.  This change disables that option to allow safe tags from user input to appear in the rendered HTML.

However, relying on `#sanitize` at the end of parsing doesn't handle javascript in tag attributes correctly.  For example, this input:

``` html
<a href="javascript:alert('foo')">link text</a>
```

is rendered as:

``` html
<a href="/namespace101/gitlabhq/blob/markdown/javascript:alert(&#39;foo&#39;)">link text</a>
```

To handle javascript in attributes, I added a call to `#sanitize` before parsing relative links.  This results in the correctly sanitized output:

``` html
<a>link text</a>
```

mr-vinn committed:
Allow HTML tags in user Markdown input

Allow whitelisted tags to appear in rendered HTML output by disabling
Redcarpet's `:filter_html` option.

mr-vinn committed:
Document whitelisted HTML tags and attributes

dblessing commented:
ping @randx 

Also, @namnatulco would you be interested in providing documentation in the meantime?

namnatulco commented:
I hope I'll find some time to do that tonight. Thanks for reminding me!

veegee commented:
What's the status of this? The gitlab docs show that HTML is possible, but  it doesn't seem to work. This patch sounds like it'll fix that.

magikid commented:
:+1:

lyfeyaj commented:
+1 I need this feature!

dosire commented:
@mr-vinn what is your plan with this?

mr-vinn commented:
@dosire I'll make the html-pipeline changes that were discussed earlier; I should be able to do that over the weekend.

dosire commented:
@mr-vinn OK, thanks!

mbroz commented:
Seems that some special formatting inside table cells is not posible with Markdown (only with added html code).
Also automatic conversion tools (from Google Code wiki for example) generate html tags inside Markdown that is ignored in gitlab (breaking pages).
Please merge support for at least some HTML tags, it increases usability a lot. Thanks!

mr-vinn committed:
Merge branch 'master' into markdown-tags

Use the latest HTML pipeline gem

mr-vinn committed:
Change HTML sanitization

Use the `SanitizationFilter` class from the html-pipeline gem for inline
HTML instead of calling the Rails `sanitize` method.

mr-vinn commented:
@dosire Sorry for the delay - I added a new change to use the html-pipeline `SanitizationFilter`, so I think this is ready for review.

dosire commented:
@randx can you review?

randx commented on lib/gitlab/markdown.rb:
```diff
@@ -79,15 +79,34 @@ def gfm_with_options(text, options = {}, project = @project, html_options = {})
       ]
 
+      whitelist = HTML::Pipeline::SanitizationFilter::WHITELIST
+      whitelist[:attributes][:all].push('class', 'id', 'style')
```
@mr-vinn I am not sure if it makes sense to allow style attributes. I dont want people override GitLab style with custom css

mr-vinn replied:
@randx That makes sense to me, I'll remove `style` from the attribute whitelist.

randx commented:
@mr-vinn please add item to CHANGELOG too

mr-vinn committed:
Don't allow style attributes in inline HTML

mr-vinn commented:
@randx The CHANGELOG and attribute whitelist have been updated in cc29ce4

bbodenmiller commented:
@me-vinn looks like you also need a commit squash & debase.

mr-vinn committed:
Merge branch 'master' into markdown-tags

Merge updated CHANGELOG entries

mr-vinn commented:
@bbodenmiller Thanks for catching that; it should be mergeable now.

randx commented:
@mr-vinn tests are failing. Can you take a look?

mr-vinn committed:
Fix SanitizationFilter bugs

Return a `SafeBuffer` instead of a `String` from the `#gfm_with_options`
method so that Rails doesn't escape our markup.

Also add `<span>` to the sanitization whitelist to avoid breaking syntax
highlighting in code blocks.

mr-vinn committed:
Merge branch 'master' into markdown-tags

mr-vinn commented:
@randx I forgot to call `#html_safe` on the parsed output, so Rails was escaping the markup.  I also had to add `<span>` tags to the sanitizer's whitelist to avoid breaking syntax highlighting in code blocks.

randx commented:
@mr-vinn ping me when it will be ready

mr-vinn committed:
Merge branch 'master' into markdown-tags

mr-vinn commented:
@randx All set, it's mergeable now.

randx commented:
@mr-vinn thank you. looks good

Razer6 commented:
@mr-vinn Nice, that you switched to html-pipeline!

zlolik commented:
Please, specify somewhere, that Ruby should have at least html-pipeline-2.0 in order to work according to documentation:
```html
<details><summary>&#9733; About html-pipeline-2.0</summary>Only that version have [merge 171 ](https://github.com/jch/html-pipeline/pull/171) that fix `<summary>` and `<details>` tags.
</details>
```
<details><summary>&#9733; About html-pipeline-2.0</summary>Only that version have [merge 171 ](https://github.com/jch/html-pipeline/pull/171) that fix `<summary>` and `<details>` tags.
</details>
